biocrnpyler/species.py

```diff
@@ -878,19 +796,461 @@ def replace(self, position, part, direction=None):
+
+        # Check whether ot make a ComplexSpecies or OrderedComplexSpecies
+        if "ordered" in keywords:
+            ordered = keywords["ordered"]
+            del keywords["ordered"]
+            ComplexClass = OrderedComplexSpecies
+        else:
+            ordered = False
+            ComplexClass = ComplexSpecies
+
+        #Use to supress errors in ComplexSpecies and OrderedComplexSpecies
+        keywords["called_from_complex"] = True
+
+        parent_species = []  # parent_species is a list of OrderedPolymerSpecies and/or PolymerConformations
+        bindlocs = []  # bindloc is the location a Species is bound to parent_species
+        insertlocs = [] #insertloc is the location of the species inside the Complex.species list (important to maintain order in OrderedComplexSpecies)
+        child_species = [] #the species with parents
+        other_species = []  # Other species in the Complex
+
+        # Below cycle through species and see if one has a parent. If it does, that means the species is
+        # in an OrderedPolymerSpecies and the Complex should be formed around it.
+        for i, specie in enumerate(species):
+            if(hasattr(specie, "parent") and (specie.parent is not None)):
+                #This adds to a growing list of parents, which will be placed in an PolymerConformation
+                # It is very important to deepcopy here because the underlying OrderedPolymerSpecies or PolymerConformation will be modified.
+                parent_species.append(specie.parent)
+                bindlocs.append(specie.position)
+                insertlocs.append(i)
+                child_species.append(specie)
+            else:
+                other_species += [specie]
+
+        #Case 1: If no OrderedPolymerSpecies is found, just call the regular constructor.
+        if len(parent_species) == 0:
+            return ComplexClass(species, *args, **keywords)
+
+        #Case 2 - 3: the Complex is being formed inside an OrderedPolymerSpecies (and only a single species has a parent).
+        elif len(parent_species) == 1 and isinstance(parent_species[0], OrderedPolymerSpecies):
+            parent_species = parent_species[0]
+            bindloc = bindlocs[0]
+
+            #Ensure the monomer in the Complex has no parent
+            #new_species = other_species
+            #monomer = copy.copy(child_species[0]) #Copy the monomer
+            #monomer.remove()
+            #new_species.insert(insertlocs[0], monomer) #Insert the monomer back into the Species
+
+            # Create an OrderedcomplexSepcies or ComplexSpecies
+            child = copy.copy(child_species[0])
+            child_direction = child_species[0].direction
+            child.remove() #Remove the child species, the new_complex will replace it
+            species.remove(child_species[0])
+            species.insert(insertlocs[0], child) #place the new child in the list
```

Comment:
took me a while to figure this out but you are doing this because you preserve the order in ordered complex. You can add a comment to explain that.

Reply:
done.